src/devices/src/virtio/balloon/device.rs

```diff
@@ -0,0 +1,920 @@
+        }
+
+        Ok(())
+    }
+}
+
+// Virtio balloon device.
+pub struct Balloon {
+    // Virtio fields.
+    pub(crate) avail_features: u64,
+    pub(crate) acked_features: u64,
+    pub(crate) config_space: ConfigSpace,
+    pub(crate) activate_evt: EventFd,
+
+    // Transport related fields.
+    pub(crate) queues: Vec<Queue>,
+    pub(crate) interrupt_status: Arc<AtomicUsize>,
+    interrupt_evt: EventFd,
+    pub(crate) queue_evts: [EventFd; NUM_QUEUES],
+    pub(crate) device_state: DeviceState,
+
+    // Implementation specific fields.
+    pub(crate) stats_polling_interval_s: u16,
+    pub(crate) stats_timer: TimerFd,
+    // The index of the previous stats descriptor is saved because
+    // it is acknowledged after the stats queue is processed.
+    pub(crate) stats_desc_index: Option<u16>,
+    pub(crate) latest_stats: BalloonStats,
+}
+
+impl Balloon {
+    pub fn new(
+        num_pages: u32,
+        must_tell_host: bool,
+        deflate_on_oom: bool,
+        stats_polling_interval_s: u16,
+    ) -> Result<Balloon, Error> {
+        let mut avail_features = 1u64 << VIRTIO_F_VERSION_1;
+
+        if must_tell_host {
+            avail_features |= 1u64 << VIRTIO_BALLOON_F_MUST_TELL_HOST;
+        };
+
+        if deflate_on_oom {
+            avail_features |= 1u64 << VIRTIO_BALLOON_F_DEFLATE_ON_OOM;
+        };
+
+        if stats_polling_interval_s > 0 {
+            avail_features |= 1u64 << VIRTIO_BALLOON_F_STATS_VQ;
+        }
+
+        let queue_evts = [
+            EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+        ];
+
+        let queues = QUEUE_SIZES.iter().map(|&s| Queue::new(s)).collect();
+
+        let stats_timer = TimerFd::new_custom(ClockId::Monotonic, true, true).map_err(Error::IO)?;
+
+        Ok(Balloon {
+            avail_features,
+            acked_features: 0u64,
+            config_space: ConfigSpace {
+                num_pages,
+                actual_pages: 0,
+            },
+            interrupt_status: Arc::new(AtomicUsize::new(0)),
+            interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            queue_evts,
+            queues,
+            device_state: DeviceState::Inactive,
+            activate_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            stats_polling_interval_s,
+            stats_timer,
+            stats_desc_index: None,
+            latest_stats: BalloonStats::default(),
+        })
+    }
+
+    pub(crate) fn process_inflate_queue_event(&mut self) {
+        if let Err(e) = self.queue_evts[INFLATE_INDEX].read() {
+            error!("Failed to get queue event: {:?}", e);
+            METRICS.balloon.event_fails.inc();
+        } else {
+            self.process_inflate()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+    }
```

Comment:
`process_inflate_queue_event()`, `process_deflate_queue_event()`, `process_stats_queue_event()` and `process_stats_timer_event()` - could be made to return `Result` and dedup failure code and hardish-to-read branching.

```suggestion
    pub(crate) fn process_inflate_queue_event(&mut self) -> Result<(), Error> {
        self.queue_evts[INFLATE_INDEX].read().map_err(Error::EventFd)?;
        self.process_inflate()
    }
```

And have the caller do `.unwrap_or_else(report_balloon_event_fail)`

Reply:
I found returning the error to the caller to be difficult as it can't use the very simple `?` mechanism. A lot of the time the error is only logged and does not stop the flow of the function in which it appears.

Furthermore, the caller (`Subscriber::process()`) also handles the activation event in the same place and the errors for these types of events are different. Even if we find a way to make the errors reach the caller, we would need to either unify the error variants of the device with the activation specific ones or treat them in separate places.

Reply:
> I found returning the error to the caller to be difficult as it can't use the very simple ? mechanism. A lot of the time the error is only logged and does not stop the flow of the function in which it appears.

Most of these `process_X_event` functions have that wild `if else if` format which can be simplified (for readability and ease of maintenance) using `?`. From what I see, the current code is logically structured to _not do other things_ after the error which is equivalent with _stopping the flow of the function_.

> Furthermore, the caller (`Subscriber::process()`) also handles the activation event in the same place and the errors for these types of events are different.

This is where you would consume or map the errors, they don't need to be all the same.
```rust
impl Subscriber for Balloon {
    fn process(self, ev: event) {
        match ev {
            X => self.process_x().unwrap_or_else(|| { /* process X err here */ }),
            Y => self.process_y().unwrap_or_else(|| { /* process Y err here */ }),
        }
    } 
}
```

To do this you would also have to clean up the `Device::Error` hierarchy a bit as it's currently a bit all over the place.

WDYT? Would it be worth it?

Reply:
I like the idea much more than how the processing happens now. I will try to refactor it a bit and see where we end up.